### PR TITLE
login logic and some small tweaks

### PR DESCRIPTION
**Pull request by rashkobby**

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -11,28 +11,30 @@ import AssignmentDetails from './pages/AssignmentDetails'
 import Certificate from './pages/Certificate'
 import PaymentSuccess from './components/Signup/PaymentSuccess'
 import ViewCertificate from './pages/ViewCertificate'
+import {useAuthContext} from './context/AuthContext'
+import { Navigate } from 'react-router-dom'
+import DataTable from './components/Dashboards/DataTable'
 
 function App() {
+  const { isAuthenticated, userRole } = useAuthContext();
```
Comment (elitekaycy) [inline src/App.jsx:19]: I think we should have a  protected route component that has the auth context logic

File: src/App.jsx [hunk 2]
```diff
@@ -11,28 +11,30 @@ import AssignmentDetails from './pages/AssignmentDetails'
...
+
   return (
     <>
       <Routes>
         <Route path="/" exact element={<Login />} />
         <Route path="/signup" element={<Signup />} />
         <Route path="/signup/redirect" element={<PaymentSuccess />} />
-        <Route path="/lessons" element={<Lessons />} />
-        <Route path="/dashboard" element={<Dashboard />}>
+        <Route path="/dashboard" element={isAuthenticated ? <Dashboard /> : <Navigate to="/" />}>
```
Comment (elitekaycy) [inline src/App.jsx:28]: this should also be in a protected route 

<ProtectedRoute>
/*
logic to switch routes to login or page depending on the auth context
*/
{children}
 </ProtectedRoute>

File: src/App.jsx (hunk 1 above)
Comment (rashkobby) [reply src/App.jsx:19]: Alright that makes sense

File: src/App.jsx (hunk 2 above)
Comment (elitekaycy) [inline src/App.jsx:28]: this should be in the protected route as well. using the auth context in a useeffect to navigate to login or render page

File: src/components/Login/LoginForm.jsx [hunk 3]
```diff
@@ -1,35 +1,43 @@
 import { useState } from 'react'
-import { Link } from 'react-router-dom'
+import { Link, useNavigate } from 'react-router-dom'
 import { BsEyeFill, BsEyeSlashFill } from 'react-icons/bs'
 import { toast } from 'react-hot-toast'
 import LoginHeading from './LoginHeading'
 import logo from '../../assets/logo.png'
+import { useAuthContext } from '../../context/AuthContext';
+
 function LoginForm() {
```
Comment (elitekaycy) [inline src/components/Login/LoginForm.jsx:9]: we should probably use react hookform for the email and password state and handle errors. its much easier

File: src/context/AuthContext.jsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+export default function AuthProvider({ children }) {
+  const [token, setToken] = useState(localStorage.getItem('token'));
+  const [userRole, setUserRole] = useState(false);
+  const navigate = useNavigate()
+  const apiUrl = import.meta.env.VITE_BACKEND_BASE_URL
+
+
+
+  const login = async (email, password) => {
```
Comment (elitekaycy) [inline src/context/AuthContext.jsx:18]: this should be in seperate loginapi.js file

File: src/context/AuthContext.jsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  const isAuthenticated = !!token;
+
+  const state = {
+    isAuthenticated,
+    userRole,
+    login,
+    logout,
+  };
+
+  const handleSubmit = async (email, password) => {
```
Comment (elitekaycy) [inline src/context/AuthContext.jsx:52]: this should also be in the loginapi.js file

File: src/context/UserContext.jsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
+import { createContext, useContext, useEffect, useState } from 'react';
```
Comment (elitekaycy) [inline src/context/UserContext.jsx:1]: the user context can be a part of the auth context to make things easier and centralized

File: src/context/UserContext.jsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      
+      fetchUserData(token);
+    } else {
+      
+      setUser(null);
+      setLoading(false);
+    }
+  }, []);
+
+  const fetchUserData = (token) => {
```
Comment (elitekaycy) [inline src/context/UserContext.jsx:27]: this could be in a helpers.js or utils

File: src/components/Login/LoginForm.jsx (hunk 3 above)
Comment (preshlele) [reply src/components/Login/LoginForm.jsx:9]: Agree, makes it simpler

File: src/App.jsx (hunk 1 above)
Comment (preshlele) [reply src/App.jsx:19]: Eslint throwing an error, the navigate import with react-router-dom are 2 atm

File: src/components/Sidebar.jsx
Comment (preshlele) [inline src/components/Sidebar.jsx:1]: The useEffect and the handleResize thing... Won't there be a better implementation ? Cos useEffect dier I dey fear re-rendering stuff

**Pull request merged**